src/lib/ecore/efl_promise2.c

```diff
@@ -0,0 +1,277 @@
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "Ecore.h"
+#include "efl_promise2.h"
+#include "ecore_private.h"
+#include "efl_promise2_priv.h"
+
+struct _Efl_Promise2 {
+   Eina_Free_Cb cancel;
+   Efl_Future2 *future;
+   const void *data;
+};
+
+struct _Efl_Future2 {
+   Efl_Future2_Cb success;
+   Efl_Future2_Cb error;
+   Eina_Free_Cb free;
+   Efl_Promise2 *promise;
+   Efl_Future2 *next;
+   Efl_Future2 *prev;
+   Eina_Value *pending_value;
+   const void *data;
+   Ecore_Event *event;
+   Eina_Bool reject;
```

Comment:
we can drop `reject`, instead `value->type == EINA_VALUE_TYPE_ERROR` means rejected.